### PR TITLE
Extending Outlining Functionality

### PR DESCRIPTION
**Pull request by cloudRoutine**
If Then Else clauses collapse
For loops and ForEach loops collapse
Hint text restricted to the range being collapsed
Abstract members collapse
... placed at the end of first line, no longer eats the =

### CODE REVIEW

File: src/FSharpVSPowerTools.Core/UntypedAstUtils.fs [hunk 1]
```diff
@@ -782,30 +782,49 @@ module Outlining =
...
+                yield! visitMatchClauses clauses
+            | SynExpr.App (_, _, _, e, _) ->
                 yield! visitExpr e
             | SynExpr.Sequential (_, _, e1, e2, _) ->
                 yield! visitExpr e1
                 yield! visitExpr e2
-            | SynExpr.YieldOrReturn (_, e, _) ->
+            | SynExpr.CompExpr (_,_,e,r) ->
+                let cexprEnd = Range.mkPos r.EndLine (r.EndColumn - 1) // exclude the closing } on the cexpr from collapsing
+                yield mkFileIndexRange r.FileIndex r.Start cexprEnd
```
Comment (dungpa) [inline src/FSharpVSPowerTools.Core/UntypedAstUtils.fs]: You could use `Range.endToEnd` function above.

File: src/FSharpVSPowerTools.Core/UntypedAstUtils.fs [hunk 2]
```diff
@@ -832,34 +851,50 @@ module Outlining =
...
+            | SynExpr.ForEach(_,_,_,_,_,e,r)->
+                yield r
+                yield! visitExpr e
+            | SynExpr.Lambda(_,_,_,e,r) ->
+                yield r
+                yield! visitExpr e
+            | SynExpr.Quote(_,isRaw,e,_,r) ->
+                // subtract columns so the @@> or @> is not collapsed
+                let rEnd = Range.mkPos r.End.Line (r.EndColumn-(if isRaw then 3 else 2)) 
+                yield mkFileIndexRange r.FileIndex r.Start rEnd
```
Comment (dungpa) [inline src/FSharpVSPowerTools.Core/UntypedAstUtils.fs]: Same comment as above.

**Comment by dungpa** [comment]
These outlining tests are failing https://github.com/fsprojects/VisualFSharpPowerTools/blob/master/tests/FSharpVSPowerTools.Core.Tests/OutliningTests.fs. Could you have a look?

**Comment by cloudRoutine** [comment]
The tests are failing because they're based on different ranges than what are currently being used. They need to be rewritten to account for the new approach.

**Comment by vasily-kirichenko** [comment]
Outlining does not work at all. I switched it in the settings on and off several times - no effect.

File: src/FSharpVSPowerTools.Logic/Outlining/OutliningTagger.fs [hunk 3]
```diff
@@ -12,107 +17,176 @@ open Microsoft.FSharp.Compiler.Ast
...
+
+[<  Export (typeof<ITaggerProvider>)
+;   ContentType "F#"
+;   TagType (typeof<IOutliningRegionTag>) >]
+type OutliningTaggerProvider [<ImportingConstructor>] 
+    ( serviceProvider: IServiceProvider
+    , textDocumentFactoryService : ITextDocumentFactoryService
+    , projectFactory: ProjectFactory
+    , vsLanguageService : VSLanguageService
+    ) as self =
```
Comment (vasily-kirichenko) [inline src/FSharpVSPowerTools.Logic/Outlining/OutliningTagger.fs]: Do not use this Haskellish style.

File: src/FSharpVSPowerTools.Logic/Outlining/OutliningTagger.fs [hunk 4]
```diff
@@ -12,107 +17,176 @@ open Microsoft.FSharp.Compiler.Ast
...
+    
+    member __.CreateTagger buffer : ITagger<IOutliningRegionTag> = 
+        let doc = ref (Unchecked.defaultof<ITextDocument>)
+        let generalOptions = Setting.getGeneralOptions (serviceProvider)
+        if  generalOptions.OutliningEnabled = true
+         && textDocumentFactoryService.TryGetTextDocument (buffer, doc) then
+            new OutliningTagger ( !doc
+                                , serviceProvider
+                                , projectFactory
+                                , vsLanguageService ) :> ITagger<_>
```
Comment (vasily-kirichenko) [inline src/FSharpVSPowerTools.Logic/Outlining/OutliningTagger.fs]: Same as above.

**Comment by vasily-kirichenko** [comment]
You added a lot of cases but no single unit test for them.

**Comment by vasily-kirichenko** [comment]
No outlining for the list

image

**Comment by vasily-kirichenko** [comment]
Great work! Thanks a lot, man.

File: src/FSharpVSPowerTools/Commands/DepthColorizerManager.cs [hunk 5]
```diff
@@ -21,28 +21,37 @@ namespace FSharpVSPowerTools
...
-        [Import]
-        internal VSLanguageService vsLanguageService = null;
+        private readonly IServiceProvider _serviceProvider;
+        private readonly ITextDocumentFactoryService _textDocumentFactoryService;
+        private readonly ProjectFactory _projectFactory;
+        private readonly VSLanguageService _vsLanguageService;
+
+        [ImportingConstructor]
+        public DepthColorizerTaggerProvider(
+            IServiceProvider serviceProvider,
```
Comment (dungpa) [inline src/FSharpVSPowerTools/Commands/DepthColorizerManager.cs]: Does this actually work? A missing `[Import(typeof(SVsServiceProvider))]` attribute means that MEF can't find the right service provider for composition.

File: src/FSharpVSPowerTools/Commands/HighlightUsageTaggerProvider.cs [hunk 6]
```diff
@@ -18,6 +18,36 @@ namespace FSharpVSPowerTools
...
+        //internal ProjectFactory projectFactory = null;
+
+        //private readonly IServiceProvider _serviceProvider;
+        //private readonly ITextDocumentFactoryService _textDocumentFactoryService;
+        //private readonly ProjectFactory _projectFactory;
+        //private readonly VSLanguageService _fsharpVsLanguageService;
+
+        //[ImportingConstructor]
+        //public HighlightUsageTaggerProvider(
+        //    IServiceProvider serviceProvider,
```
Comment (dungpa) [inline src/FSharpVSPowerTools/Commands/HighlightUsageTaggerProvider.cs]: If we add `[Import(typeof(SVsServiceProvider))]`, we can go with `ImportingConstructor` refactoring.

**Pull request merged**